src/librustc_codegen_ironox/metadata.rs

```diff
@@ -31,8 +32,30 @@ impl MetadataLoader for IronOxMetadataLoader {
         Ok(rustc_erase_owner!(buf))
     }
 
-    fn get_dylib_metadata(&self, _target: &Target, filename: &Path)
+    fn get_dylib_metadata(&self, target: &Target, filename: &Path)
         -> Result<MetadataRef, String> {
-        unimplemented!("get_dylib_metadata {:?}", filename);
+        let mut input_file =
+            File::open(filename)
+                .map_err(|_e| format!("failed to read {}", filename.display()))?;
+        let mut buf = vec![];
+        input_file.read_to_end(&mut buf)
+            .map_err(|_| format!("failed to read {}", filename.display()))?;
+        let buf = OwningRef::new(box buf);
+        let buf = buf.try_map(|buf| search_meta_section(&buf, target, filename))?;
+        return Ok(rustc_erase_owner!(buf));
+    }
+}
+
+fn search_meta_section<'a>(bytes: &'a [u8], target: &Target, filename: &Path)
+    -> Result<&'a [u8], String> {
+    let elf = goblin::elf::Elf::parse(&bytes).map_err(|_| "failed to parse ELF")?;
+    for sh in &elf.section_headers {
+        if elf.shdr_strtab.get(sh.sh_name)
+            .map_or(false, |r| r.ok() == Some(metadata_section_name(target))) {
```

Comment:
Could this be something like:

```rust
if elf.sections_headers.iter().find(|x| if let Ok(Some(metadata_section_name(target) = elf.shdr_strtab.get(x.sh_name)) { ... }
```

Reply:
What Laurie said.

I'm a bit surprised goblin doesn't have the ability to find a section by name (it doesn't, I checked).

Reply:
`Ok(Some(metadata_section_name(target)))` wouldn't work, because the error of the result returned by `get` is a `goblin:error:Error`.

Reply:
Would `map_err` on the right-hand side help?

Reply:
OK.

Reply:
I needed `map_or`, since `get` returns an `Option<Result<...>>`. The best I could come up with is:
```
.find(|x| Some(metadata_section_name(target)) == elf.shdr_strtab.get(x.sh_name).map_or(None, |x| x.ok()));
```